File: Server/src/tools/measurement.py
"""Measurement tools for Fusion 360 MCP Server.

Contains functions for measuring distances, angles, areas, volumes, and other geometric properties.
Uses @fusion_tool decorator for automatic HTTP handling and telemetry.
"""

from .base import fusion_tool


@fusion_tool
def measure_distance(
    entity1_type: str,
    entity1_index: int,
    entity2_type: str,
    entity2_index: int,
    body1_index: int = 0,
    body2_index: int = 0,
):
    """
    Measure the minimum distance between two entities.

    Args:
        entity1_type: Type of first entity ("face", "edge", "vertex", "body", "point")
        entity1_index: Index of first entity (for face/edge/vertex within the body)
        entity2_type: Type of second entity ("face", "edge", "vertex", "body", "point")
        entity2_index: Index of second entity (for face/edge/vertex within the body)
        body1_index: Index of the body containing entity1 (default: 0)
        body2_index: Index of the body containing entity2 (default: 0)

    Returns:
        Dictionary with:
        - distance: Minimum distance in cm
        - point1: [x, y, z] closest point on entity1
        - point2: [x, y, z] closest point on entity2
    """


@fusion_tool
def measure_angle(
    entity1_type: str,
    entity1_index: int,
    entity2_type: str,
    entity2_index: int,
    body1_index: int = 0,
    body2_index: int = 0,
):
    """
    Measure the angle between two planar faces or linear edges.

    Args:
        entity1_type: Type of first entity ("face" or "edge")
        entity1_index: Index of first entity
        entity2_type: Type of second entity ("face" or "edge")
        entity2_index: Index of second entity
        body1_index: Index of the body containing entity1 (default: 0)
        body2_index: Index of the body containing entity2 (default: 0)

    Returns:
        Dictionary with:
        - angle_degrees: Angle in degrees
        - angle_radians: Angle in radians
    """


@fusion_tool
def measure_area(face_index: int, body_index: int = 0):
    """
    Measure the area of a specific face.

    Args:
        face_index: Index of the face to measure
        body_index: Index of the body containing the face (default: 0)

    Returns:
        Dictionary with:
        - area_cm2: Area in square centimeters
        - area_mm2: Area in square millimeters
        - face_type: Type of face (Plane, Cylinder, etc.)
    """


@fusion_tool
def measure_volume(body_index: int = 0):
    """
    Measure the volume of a body.

    Args:
        body_index: Index of the body to measure (default: 0)

    Returns:
        Dictionary with:
        - volume_cm3: Volume in cubic centimeters
        - volume_mm3: Volume in cubic millimeters
        - body_name: Name of the body
    """


@fusion_tool
def measure_edge_length(edge_index: int, body_index: int = 0):
    """
    Measure the length of a specific edge.

    Args:
        edge_index: Index of the edge to measure
        body_index: Index of the body containing the edge (default: 0)

    Returns:
        Dictionary with:
        - length_cm: Length in centimeters
        - length_mm: Length in millimeters
        - edge_type: Type of edge (Line, Arc, Circle, etc.)
        - start_point: [x, y, z] start point
        - end_point: [x, y, z] end point
    """


@fusion_tool
def measure_body_properties(body_index: int = 0):
    """
    Get comprehensive physical properties of a body.

    Args:
        body_index: Index of the body to measure (default: 0)

    Returns:
        Dictionary with:
        - volume_cm3: Volume in cubic centimeters
        - surface_area_cm2: Total surface area in square centimeters
        - bounding_box: {min: [x,y,z], max: [x,y,z], size: [dx,dy,dz]}
        - centroid: [x, y, z] center of mass
        - face_count: Number of faces
        - edge_count: Number of edges
        - vertex_count: Number of vertices
        - body_name: Name of the body
    """


@fusion_tool
def measure_point_to_point(point1: list, point2: list):
    """
    Measure the distance between two specific 3D points.

    Args:
        point1: [x, y, z] coordinates of first point (in cm)
        point2: [x, y, z] coordinates of second point (in cm)

    Returns:
        Dictionary with:
        - distance_cm: Distance in centimeters
        - distance_mm: Distance in millimeters
        - delta: [dx, dy, dz] difference vector
    """


@fusion_tool(method="GET")
def get_edges_info(body_index: int = 0):
    """
    Get detailed edge information for a body.

    Args:
        body_index: Index of the body to inspect (default: 0)

    Returns:
        Dictionary with:
        - edge_count: Number of edges
        - edges: List with details for each edge (index, type, length, start_point, end_point)
    """


@fusion_tool(method="GET")
def get_vertices_info(body_index: int = 0):
    """
    Get detailed vertex information for a body.

    Args:
        body_index: Index of the body to inspect (default: 0)

    Returns:
        Dictionary with:
        - vertex_count: Number of vertices
        - vertices: List with details for each vertex (index, position [x,y,z])
    """

File: Server-TS/src/tools/measurement.ts
/**
 * Measurement tools for Fusion 360 MCP Server.
 */

import { z } from "zod";
import { createPostTool, createGetTool, ToolResult } from "./base.js";

// Zod schemas
export const measureDistanceSchema = z.object({
  entity1_type: z
    .string()
    .describe("Type: 'face', 'edge', 'vertex', or 'body'"),
  entity1_index: z.number().describe("Index of first entity"),
  entity2_type: z
    .string()
    .describe("Type: 'face', 'edge', 'vertex', or 'body'"),
  entity2_index: z.number().describe("Index of second entity"),
  body1_index: z.number().default(0).describe("Body index for first entity"),
  body2_index: z.number().default(0).describe("Body index for second entity"),
});

export const measureAngleSchema = z.object({
  entity1_type: z.string().describe("Type: 'face' or 'edge'"),
  entity1_index: z.number().describe("Index of first entity"),
  entity2_type: z.string().describe("Type: 'face' or 'edge'"),
  entity2_index: z.number().describe("Index of second entity"),
  body1_index: z.number().default(0).describe("Body index for first entity"),
  body2_index: z.number().default(0).describe("Body index for second entity"),
});

export const measureAreaSchema = z.object({
  face_index: z.number().describe("Index of the face"),
  body_index: z.number().default(0).describe("Index of the body"),
});

export const measureVolumeSchema = z.object({
  body_index: z.number().default(0).describe("Index of the body"),
});

export const measureEdgeLengthSchema = z.object({
  edge_index: z.number().describe("Index of the edge"),
  body_index: z.number().default(0).describe("Index of the body"),
});

export const measureBodyPropertiesSchema = z.object({
  body_index: z.number().default(0).describe("Index of the body"),
});

export const measurePointToPointSchema = z.object({
  point1: z.array(z.number()).length(3).describe("[x, y, z] of first point"),
  point2: z.array(z.number()).length(3).describe("[x, y, z] of second point"),
});

export const getEdgesInfoSchema = z.object({
  body_index: z.number().default(0).describe("Index of the body to inspect"),
});

export const getVerticesInfoSchema = z.object({
  body_index: z.number().default(0).describe("Index of the body to inspect"),
});

// Tool implementations
const _measureDistance = createPostTool("measure_distance");
const _measureAngle = createPostTool("measure_angle");
const _measureArea = createPostTool("measure_area");
const _measureVolume = createPostTool("measure_volume");
const _measureEdgeLength = createPostTool("measure_edge_length");
const _measureBodyProperties = createPostTool("measure_body_properties");
const _measurePointToPoint = createPostTool("measure_point_to_point");
const _getEdgesInfo = createGetTool("edges_info");
const _getVerticesInfo = createGetTool("vertices_info");

// Exported functions

/**
 * Measure the minimum distance between two entities.
 */
export async function measure_distance(
  entity1_type: string,
  entity1_index: number,
  entity2_type: string,
  entity2_index: number,
  body1_index: number = 0,
  body2_index: number = 0
): Promise<ToolResult> {
  return _measureDistance({
    entity1_type,
    entity1_index,
    entity2_type,
    entity2_index,
    body1_index,
    body2_index,
  });
}

/**
 * Measure the angle between two planar faces or linear edges.
 */
export async function measure_angle(
  entity1_type: string,
  entity1_index: number,
  entity2_type: string,
  entity2_index: number,
  body1_index: number = 0,
  body2_index: number = 0
): Promise<ToolResult> {
  return _measureAngle({
    entity1_type,
    entity1_index,
    entity2_type,
    entity2_index,
    body1_index,
    body2_index,
  });
}

/**
 * Measure the area of a specific face.
 */
export async function measure_area(
  face_index: number,
  body_index: number = 0
): Promise<ToolResult> {
  return _measureArea({ face_index, body_index });
}

/**
 * Measure the volume of a body.
 */
export async function measure_volume(
  body_index: number = 0
): Promise<ToolResult> {
  return _measureVolume({ body_index });
}

/**
 * Measure the length of a specific edge.
 */
export async function measure_edge_length(
  edge_index: number,
  body_index: number = 0
): Promise<ToolResult> {
  return _measureEdgeLength({ edge_index, body_index });
}

/**
 * Get comprehensive physical properties of a body.
 */
export async function measure_body_properties(
  body_index: number = 0
): Promise<ToolResult> {
  return _measureBodyProperties({ body_index });
}

/**
 * Measure the distance between two specific 3D points.
 */
export async function measure_point_to_point(
  point1: number[],
  point2: number[]
): Promise<ToolResult> {
  return _measurePointToPoint({ point1, point2 });
}

/**
 * Get detailed edge information for a body.
 */
export async function get_edges_info(
  body_index: number = 0
): Promise<ToolResult> {
  return _getEdgesInfo({ body_index });
}

/**
 * Get detailed vertex information for a body.
 */
export async function get_vertices_info(
  body_index: number = 0
): Promise<ToolResult> {
  return _getVerticesInfo({ body_index });
}
